---
title: "RNASeq Report"
author:
- Paul Cao
output: pdf_document
fontsize: 12pt
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, echo=FALSE, message=FALSE}
library("tximport")
library("readr")
library("tximportData")
dir <- system.file("extdata", package="tximportData")
```

## File Inputs

### Samples Sheet
```{r, echo=FALSE, message=FALSE}
samples2 <- read.table(file.path("/", "samples.txt"), head=TRUE)
samples2
```

### Salmon Quantification Inputs

```{r, echo=FALSE, message=FALSE}
files <- file.path("/", samples2$Sample, "quant.sf")
files
```


## Differential expression analysis

### DESeq2 Summary

The standard differential expression analysis steps are wrapped into a single function, DESeq. The estimation steps performed by this function are described below, in the manual page for ?DESeq and in the Methods section of the DESeq2 publication (Love, Huber, and Anders 2014).

#### Summary Statistics

```{r, echo=FALSE, message=FALSE, warning=FALSE}
library("DESeq2")

tx2gene <- read_csv(file.path(dir, "tx2gene.gencode.v27.csv"))
txi <- tximport(files, type="salmon", tx2gene=tx2gene)

dds <- DESeqDataSetFromTximport(txi, colData = samples2, design = ~ Condition)
dds <- DESeq(dds)
res <- results(dds)
summary(res)
```

#### DESeq2 Raw Result

```{r, echo=FALSE, message=FALSE, warning=FALSE}
res
```

#### Log fold change shrinkage for visualization and ranking

Shrinkage of effect size (LFC estimates) is useful for visualization and ranking of genes. To shrink the LFC, we pass the dds object to the function lfcShrink. Below we specify to use the apeglm method for effect size shrinkage (Zhu, Ibrahim, and Love 2018), which improves on the previous estimator.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
res <- results(dds, name="Condition_B_vs_A")
res <- results(dds, contrast=c("Condition","B","A"))

resLFC <- lfcShrink(dds, coef="Condition_B_vs_A", type="apeglm")
resLFC
```

### Exploring and exporting results

#### MA-plot

In DESeq2, the function plotMA shows the log2 fold changes attributable to a given variable over the mean of normalized counts for all the samples in the DESeqDataSet. Points will be colored red if the adjusted p value is less than 0.1. Points which fall out of the window are plotted as open triangles pointing either up or down.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
plotMA(res, ylim=c(-2,2))
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
plotMA(resLFC, ylim=c(-2,2))
```

#### Alternative shrinkage estimators

The moderated log fold changes proposed by Love, Huber, and Anders (2014) use a normal prior distribution, centered on zero and with a scale that is fit to the data. The shrunken log fold changes are useful for ranking and visualization, without the need for arbitrary filters on low count genes. The normal prior can sometimes produce too strong of shrinkage for certain datasets. In DESeq2 version 1.18, we include two additional adaptive shrinkage estimators, available via the type argument of lfcShrink. For more details, see ?lfcShrink

The options for type are:

    apeglm is the adaptive t prior shrinkage estimator from the apeglm package (Zhu, Ibrahim, and Love 2018). As of version 1.28.0, it is the default estimator.
    ashr is the adaptive shrinkage estimator from the ashr package (Stephens 2016). Here DESeq2 uses the ashr option to fit a mixture of Normal distributions to form the prior, with method="shrinkage".
    normal is the the original DESeq2 shrinkage estimator, an adaptive Normal distribution as prior.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
resNorm <- lfcShrink(dds, coef=2, type="normal")
resAsh <- lfcShrink(dds, coef=2, type="ashr")

par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
plotMA(resLFC, xlim=xlim, ylim=ylim, main="apeglm")
plotMA(resNorm, xlim=xlim, ylim=ylim, main="normal")
plotMA(resAsh, xlim=xlim, ylim=ylim, main="ashr")
```


### Data quality assessment by sample clustering and visualization

Data quality assessment and quality control (i.e. the removal of insufficiently good data) are essential steps of any data analysis. These steps should typically be performed very early in the analysis of a new data set, preceding or in parallel to the differential expression testing.

We define the term quality as fitness for purpose. Our purpose is the detection of differentially expressed genes, and we are looking in particular for samples whose experimental treatment suffered from an anormality that renders the data points obtained from these particular samples detrimental to our purpose.


#### Heatmap of the count matrix

To explore a count matrix, it is often instructive to look at it as a heatmap. Below we show how to produce such a heatmap for various transformations of the data.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
library("pheatmap")
ntd <- normTransform(dds)
select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:20]
df <- as.data.frame(colData(dds)[,c("Condition")])
pheatmap(assay(ntd)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)
```